src/index/coinstatsindex.h

```diff
@@ -49,7 +51,11 @@ class CoinStatsIndex final : public BaseIndex
 
     BaseIndex::DB& GetDB() const override { return *m_db; }
 
-    const char* GetName() const override { return "coinstatsindex"; }
+    const std::string& GetName() const override
+    {
+        static std::string name{"coinstatsindex"};
+        return name;
+    }
```

Comment:
Can the temporary here (and in txindex.h) be omitted? (also, why `static`)

Reply:
We need to return a `const std::string&` reference, so the underlying object can't be destroyed after the function returns, which is achieved with the `static`. I don't know of a better way to implement this but I'm open to suggestions?

Reply:
Ah, right.

Reply:
No strong feelings, but I'm not sure it's an improvement:
- doesn't reduce the amount of code (disregarding newlines)
- moves the name further away from the only place where it's used, so imo slightly harder to read
- requires index_name to be initialized for every instance of the class instead of just once

So with that in mind, I think I'd rather keep it as is?

Reply:
It may also be possible to return a string_view (from a lifetime perspective this is no different than a reference), which works for strings and char-string literals.

string_view should work with tinyformat as well as std::string::operator+

Reply:
<details><summary>Another approach :smile: </summary><p>

```diff
diff --git a/src/index/coinstatsindex.cpp b/src/index/coinstatsindex.cpp
index b9029e946a..a554cd4e3b 100644
--- a/src/index/coinstatsindex.cpp
+++ b/src/index/coinstatsindex.cpp
@@ -23,6 +23,7 @@ using node::UndoReadFromDisk;
 static constexpr uint8_t DB_MUHASH{'M'};
+static const std::string index_name{"coinstatsindex"};
 
@@ -392,6 +393,8 @@ bool CoinStatsIndex::CustomInit(const std::optional<interfaces::BlockKey>& block
 
+const std::string& CoinStatsIndex::GetName() const { return index_name; }
+
 bool CoinStatsIndex::CustomCommit(CDBBatch& batch)
diff --git a/src/index/coinstatsindex.h b/src/index/coinstatsindex.h
index 91a30fb9d9..54fff61e53 100644
--- a/src/index/coinstatsindex.h
+++ b/src/index/coinstatsindex.h
@@ -51,11 +51,7 @@ protected:
 
-    const std::string& GetName() const override
-    {
-        static std::string name{"coinstatsindex"};
-        return name;
-    }
+    const std::string& GetName() const override;
 
 public:
     // Constructs the index, which becomes available to be queried.
diff --git a/src/index/txindex.cpp b/src/index/txindex.cpp
index b719aface8..489bcd53c0 100644
--- a/src/index/txindex.cpp
+++ b/src/index/txindex.cpp
@@ -12,6 +12,7 @@
 constexpr uint8_t DB_TXINDEX{'t'};
+static const std::string index_name{"txindex"};
  
@@ -54,6 +55,8 @@ TxIndex::TxIndex(std::unique_ptr<interfaces::Chain> chain, size_t n_cache_size,
 
+const std::string& TxIndex::GetName() const { return index_name; }
+
 bool TxIndex::CustomAppend(const interfaces::BlockInfo& block)
 {
     // Exclude genesis block transaction because outputs are not spendable.
diff --git a/src/index/txindex.h b/src/index/txindex.h
index cf0ecf498f..c54ad562d8 100644
--- a/src/index/txindex.h
+++ b/src/index/txindex.h
@@ -29,11 +29,7 @@ protected:
 
-    const std::string& GetName() const override
-    {
-        static std::string name{"txindex"};
-        return name;
-    }
+    const std::string& GetName() const override;
 
 public:
 ``
</p></details>

Reply:
> Edit: to clarify, for example in util::ThreadRename, when passed a string_view you wouldn't be able to concatenate "b-" + name, would have to be "b-" + name.data() 

`"b-" + name.data()` is undefined behaviour if it compiles.

I was thinking that you'd be able to write `"b-"s + name`, but I was wrong.

Reply:
Thank you for the input everyone, much appreciated!

I've incorporated [sipa's suggestion](https://github.com/bitcoin/bitcoin/pull/25971#issuecomment-1234301843) to move the name logic to `BaseIndex`, which I think is an elegant solution and should hopefully also satisfy jonatack's concerns? The only downside to this approach I see is that there is some [code duplication](https://github.com/bitcoin/bitcoin/compare/4f7a2b64d8398238fc364fee9f5f2bcc4832db77..a6927e89937fdaedbc42a243dea4c90b24519378#diff-bee14d2136fca480f898cfe34d606ffb7c7460888bfb7654c575c39a77ca00ddR100-R104) in the BlockFilterIndex constructor as we need to derive the filter name from the filter type, but it's minimal enough imo.

I've also updated the thread name parameters to `string_view` wherever it made sense.

> so you could write...

Thanks for the clarification! The call to `SetInternalName()` on the next line also requires a string, so I think it's best to just expect a `string` type as parameter (reverting it to what it was before).

---

I _think_ this addresses everyone's concerns voiced here, but I'll leave the comment open for a bit longer so people can react if I misunderstood anything.

Reply:
Closing this thread as resolved